### ISSUE TITLE
Corruptions with concurent lazy

### ISSUE THREAD

**Issue by chambart** [ocaml-multicore/ocaml-multicore#329]
I don't know if #281 should have fixed this, but I have some problem triggering both with the master (d947bc7e72a0597706194ce5bdc257b0f065c989) and parallel_minor (4218b0e382af4c5bb39c5f3124808e060fb79943) branches.

```OCaml
let f count =
  let n = (Domain.self ():> int) in
  Printf.printf "forcing: %i\n%!" n;
  let r = ref 0 in
  for i = 1 to count do
    incr r;
  done;
  !r

let l = lazy (f 1_000_000_000)
let d1 =
  Domain.spawn (fun () ->
      let n = (Domain.self ():> int) in
      (* Required for the failure, probably to delay *)
      Printf.printf "spawned: %i\n%!" n;
      Lazy.force l)
let n2 = Lazy.force l
let n1 = Domain.join d1

let () = Printf.printf "forced: %i %i\n%!" n1 n2
```
The result is usually something like
```
spawned: 1
forcing: 0
forced: 70116099571644 1000000000
```

Note that the corruption doesn't happen when the domain running the force is the domain 0

**Comment by kayceesrk** [comment]
There was a missing case for handling lazy value forcing in `matching.ml`. #330 fixes it. 

Since there is a concurrent forcing of lazy, `Lazy.Undefined` is raised.

```ocaml
(* lazy1.ml *)
let f count =
  let n = (Domain.self ():> int) in
  Printf.printf "forcing on domain %d\n%!" n;
  let r = ref 0 in
  for i = 1 to count do
    incr r;
  done;
  !r

let l = lazy (f 1_000_000_000)
let d1 =
  Domain.spawn (fun () ->
      let n = (Domain.self ():> int) in
      Printf.printf "spawned domain %d\n%!" n;
      Lazy.force l)
let n2 = Lazy.force l
let n1 = Domain.join d1

let _ = Printf.printf "n1=%d n2=%d\n" n1 n2
```

```bash
$ ./lazy1.exe
spawned domain 1
forcing on domain 0
Fatal error: exception CamlinternalLazy.Undefined
```

You can handle concurrent forcing of lazy like so:

```ocaml
(* lazy2 *)
(* Not safe if lazy forces itself; spins infinitely *)
let rec quasi_safe_force l =
  try Lazy.force l with
  | Lazy.Undefined ->
      Domain.Sync.cpu_relax ();
      quasi_safe_force l

let f count =
  let n = (Domain.self ():> int) in
  Printf.printf "forcing on domain %d\n%!" n;
  let r = ref 0 in
  for i = 1 to count do
    incr r;
  done;
  !r

let l = lazy (f 1_000_000_000)
let d1 =
  Domain.spawn (fun () ->
      let n = (Domain.self ():> int) in
      Printf.printf "spawned domain %d\n%!" n;
      quasi_safe_force l)
let n2 = quasi_safe_force l
let n1 = Domain.join d1

let _ = Printf.printf "n1=%d n2=%d\n" n1 n2
```

```bash
spawned domain 1
forcing on domain 0
n1=1000000000 n2=1000000000
```

**Comment by kayceesrk** [comment]
Fixed by #330.
